src/factories/createInternalHttpTerminator.ts

```diff
@@ -131,16 +131,21 @@ export default (
       destroySocket(socket);
     }
 
-    if (sockets.size) {
-      await delay(configuration.gracefulTerminationTimeout);
-
+    // Wait for all in-flight connections to drain, forcefully terminating any
+    // open connections after the given timeout
+    try {
+      await waitFor(() => {
+        return sockets.size === 0 && secureSockets.size === 0;
+      }, {
+        interval: 1,
```

Comment:
Note that `p-wait-for` polls for this condition every 1ms. This feels a little janky, but was the most straightforward solution to me (and clearly an improvement over the current behavior). Open to alternative suggestions if this also feels gross to you.